server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -2773,4 +2776,55 @@ private static void trimUnsafeCommits(EngineConfig engineConfig) throws IOExcept
+     * after the local checkpoint in the safe commit. This step ensures the live version map and checkpoint tracker
+     * in sync with Lucene commit.
+     */
+    private void restoreVersionMapAndCheckpointTracker(DirectoryReader directoryReader) throws IOException {
+        final IndexSearcher searcher = new IndexSearcher(directoryReader);
+        searcher.setQueryCache(null);
+        final Query query = LongPoint.newRangeQuery(SeqNoFieldMapper.NAME, getLocalCheckpoint() + 1, Long.MAX_VALUE);
+        final Weight weight = searcher.createWeight(query, ScoreMode.COMPLETE_NO_SCORES, 1.0f);
+        for (LeafReaderContext leaf : directoryReader.leaves()) {
+            final Scorer scorer = weight.scorer(leaf);
+            if (scorer == null) {
+                continue;
+            }
+            final CombinedDocValues dv = new CombinedDocValues(leaf.reader());
+            final DocIdSetIterator iterator = scorer.iterator();
+            int docId;
+            while ((docId = iterator.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
+                final long primaryTerm = dv.docPrimaryTerm(docId);
+                if (primaryTerm == -1L) {
+                    continue; // skip children docs which do not have primary term
+                }
+                final long seqNo = dv.docSeqNo(docId);
+                localCheckpointTracker.markSeqNoAsCompleted(seqNo);
+                final IdOnlyFieldVisitor idFieldVisitor = new IdOnlyFieldVisitor();
+                leaf.reader().document(docId, idFieldVisitor);
+                if (idFieldVisitor.getId() == null) {
+                    assert dv.isTombstone(docId);
+                    continue;
+                }
+                final BytesRef uid = new Term(IdFieldMapper.NAME, Uid.encodeId(idFieldVisitor.getId())).bytes();
+                try (Releasable ignored = versionMap.acquireLock(uid)) {
+                    final VersionValue curr = versionMap.getUnderLock(uid);
+                    if (curr == null ||
+                        compareOpToVersionMapOnSeqNo(idFieldVisitor.getId(), seqNo, primaryTerm, curr) == OpVsLuceneDocStatus.OP_NEWER) {
+                        if (dv.isTombstone(docId)) {
+                            // use the current timestamp to honor GC deletes in case this engine becomes primary.
+                            final long startTime = config().getThreadPool().relativeTimeInMillis();
+                            versionMap.putDeleteUnderLock(uid, new DeleteVersionValue(dv.docVersion(docId), seqNo, primaryTerm, startTime));
+                        } else {
+                            versionMap.putIndexUnderLock(uid, new IndexVersionValue(null, dv.docVersion(docId), seqNo, primaryTerm));
+                        }
+                    }
+                }
+            }
+        }
+        // remove live entries in the version map
+        refresh("restore_version_map_and_checkpoint_tracker", SearcherScope.INTERNAL, true);
```

Comment:
Isn't this just calling `maybePruneDeletes`, which seems to only run if X amount of time has passed after the last pruning? Isn't this a problem in general if we want to mainly do pruning based on local checkpoint?

Reply:
We can manually call `versionMap.pruneTombstones(Long.MAX_VALUE, getLocalCheckpoint());` after the refresh. However, I expect it won't prune anything since all delete tombstones we have should be above the local checkpoint.

Reply:
The local checkpoint in the commit is a conservative approximation of the the true local checkpoint, but you're right, this should not lead to much pruning here. I'm generally concerned with the `maybePruneDeletes` method, which only seems to care about timestamps, and not local checkpoint advancement. Not something we need to solve in this PR, just a more general observation.